File: src/main/java/me/zeroeightsix/kami/module/modules/player/YawLock.java
package me.zeroeightsix.kami.module.modules.player;

import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import net.minecraft.util.math.MathHelper;

/**
 * Created by 086 on 16/12/2017.
 */
@Module.Info(name = "YawLock", category = Module.Category.PLAYER, description = "Locks your camera yaw")
public class YawLock extends Module {
    private Setting<Boolean> auto = register(Settings.b("Auto", true));
    private Setting<Float> yaw = register(Settings.f("Yaw", 180));
    private Setting<Integer> slice = register(Settings.i("Slice", 8));

    @Override
    public void onUpdate() {
        if (slice.getValue() == 0) return;
        if (auto.getValue()) {
            int angle = 360 / slice.getValue();
            float yaw = mc.player.rotationYaw;
            yaw = Math.round(yaw / angle) * angle;
            mc.player.rotationYaw = yaw;
            if (mc.player.isRiding()) mc.player.getRidingEntity().rotationYaw = yaw;
        } else {
            mc.player.rotationYaw = MathHelper.clamp(yaw.getValue() - 180, -180, 180);
        }
    }
}

File: src/main/java/me/zeroeightsix/kami/module/modules/player/YawLock.kt
package me.zeroeightsix.kami.module.modules.player

import me.zeroeightsix.kami.event.events.SafeTickEvent
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.Settings
import me.zeroeightsix.kami.util.event.listener
import kotlin.math.round

@Module.Info(
        name = "YawLock",
        category = Module.Category.PLAYER,
        description = "Locks your camera yaw"
)
object YawLock : Module() {
    private val auto = register(Settings.b("Auto", true))
    private val yaw = register(Settings.floatBuilder("Yaw").withValue(180.0f).withRange(-180.0f, 180.0f).withStep(1.0f))
    private val slice = register(Settings.integerBuilder("Slice").withValue(8).withRange(2, 32).withStep(1).withVisibility { auto.value })

    init {
        listener<SafeTickEvent> {
            if (auto.value) {
                val angle = 360.0f / slice.value
                mc.player.rotationYaw = round(mc.player.rotationYaw / angle) * angle
                mc.player.ridingEntity?.let { it.rotationYaw = mc.player.rotationYaw }
            } else {
                mc.player.rotationYaw = yaw.value
            }
        }
    }
}